iceoryx_posh/source/popo/waitset/wait_set.cpp

```diff
@@ -0,0 +1,127 @@
+// limitations under the License
+
+#include "iceoryx_posh/internal/popo/waitset/wait_set.hpp"
+
+namespace iox
+{
+namespace popo
+{
+WaitSet::WaitSet(cxx::not_null<ConditionVariableData* const> condVarDataPtr) noexcept
+    : m_conditionVariableDataPtr(condVarDataPtr)
+    , m_conditionVariableWaiter(m_conditionVariableDataPtr)
+    , m_guardCondition(m_conditionVariableDataPtr)
+{
+    m_conditionVector.push_back(&m_guardCondition);
+}
+
+bool WaitSet::attachCondition(Condition& condition) noexcept
+{
+    if (!condition.isConditionVariableAttached())
+    {
+        if (condition.attachConditionVariable(m_conditionVariableDataPtr))
+        {
+            return m_conditionVector.push_back(&condition);
+        }
+    }
+
+    return false;
+}
+
+bool WaitSet::detachCondition(Condition& condition) noexcept
+{
+    for (auto currentCondition : m_conditionVector)
+    {
+        if (currentCondition == &condition)
+        {
+            m_conditionVector.erase(&currentCondition);
+            return true;
+        }
+    }
+    return false;
+}
+
+void WaitSet::clear() noexcept
+{
+    m_conditionVector.clear();
+}
+
+cxx::vector<Condition*, MAX_NUMBER_OF_CONDITIONS>
+WaitSet::waitAndReturnFulfilledConditions(bool enableTimeout, units::Duration timeout) noexcept
+{
+    cxx::vector<Condition*, MAX_NUMBER_OF_CONDITIONS> conditionsWithFulfilledPredicate;
+
+    /// @note Inbetween here and last wait someone could have set the trigger to true, hence reset it
+    m_conditionVariableWaiter.reset();
+
+    // Is one of the conditons true?
+    for (auto currentCondition : m_conditionVector)
+    {
+        if (currentCondition->hasTrigger())
+        {
+            conditionsWithFulfilledPredicate.push_back(currentCondition);
+            currentCondition->resetTrigger();
+        }
+    }
+
+    if (conditionsWithFulfilledPredicate.empty())
+    {
+        if (enableTimeout)
+        {
+            auto retVal = m_conditionVariableWaiter.timedWait(timeout);
+
+            if (retVal == false)
+            {
+                // Return empty list
+                return conditionsWithFulfilledPredicate;
+            }
+        }
+        else
+        {
+            m_conditionVariableWaiter.wait();
+        }
+
+        // Check again if one of the conditions is true after we received the signal
+        for (auto currentCondition : m_conditionVector)
+        {
+            if (currentCondition->hasTrigger())
+            {
+                conditionsWithFulfilledPredicate.push_back(currentCondition);
+                currentCondition->resetTrigger();
+            }
+        }
+    }
+    // Return of a copy of all conditions that were fulfilled
+    return conditionsWithFulfilledPredicate;
+}
+
+cxx::vector<Condition*, MAX_NUMBER_OF_CONDITIONS> WaitSet::timedWait(units::Duration timeout) noexcept
+{
+    bool enableTimeout{true};
+    return waitAndReturnFulfilledConditions(enableTimeout, timeout);
```

Comment:
This is an example why I'm against boolean parameter in non setter/enabler methods. You have to create additional variables to make it clear what the parameter means and each user of this function needs to resist the temptation to just pass `true` or `false` to it.
If you want to keep the bool, please make `enableTimeout` a constexpr

Reply:
Refactored with a `WaitPolicy`